### ISSUE TITLE
provide Medium-style email authentication

### ISSUE THREAD

**Issue by whit537** [gratipay/gratipay.com#1052]
We should provide [Medium-style email authentication](https://github.com/gratipay/gratipay.com/issues/1052#issuecomment-125301724).
### Old

We should provide password authentication. You should be able to use either your username and/or an email address (which one? only primary? configurable per-address?). This ticket started out being about signing up with email, but the actual important/hard part is the password authentication.

We first discussed email sign-up on #89, but that became more about linking email accounts to existing Gittip accounts rather than using email to sign up in the first place. Here's a ticket about signing up with email in the first place, in response to a request I received in private email:

> Is there a way to give money without signing up for Twitter, etc.?

We currently rely on social network profiles to detect fraud. It does limit our reach, however.
#### Notify:
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2358
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/2749
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/3213
- [ ] https://gratipay.freshdesk.com/helpdesk/tickets/6349
##

**Comment by whit537** [comment]
Some recent thinking on design for this:

http://blog.codinghorror.com/the-god-login/

**Comment by rummik** [comment]
@whit537 I'd also recommend trying to help break the trend that [xkcd 936](http://xkcd.com/936/) illustrates by giving an example using multiple words, and possibly by pushing the term 'passphrase'.

Also, that was a good read :P  Not sure why I hadn't thought of making login attempts progressively slower as more failures occur, since that's something that's been done by Unix and Unix-likes for decades.

**Comment by slaivyn** [comment]
+1 and thx for "the God login", very interesting!

**Comment by whit537** [comment]
+1 https://gratipay.freshdesk.com/helpdesk/tickets/1667

**Comment by whit537** [comment]
[Something to consider](https://twitter.com/homakov/status/577269918296633344) when we get around to implementing this:

> Attack on Rails apps: sign up as victim@gmail.com, copy _app_sess, wait for the victim to set a new password, then use saved session

**Comment by seanlinsley** [comment]
Hah, did you see my RT of that? :stuck_out_tongue_winking_eye: 

The vulnerability being that Rails doesn't invalidate session tokens stored as cookies, and that the most popular authentication library for Rails (Devise) doesn't require email confirmation by default.

That wouldn't be an issue for Gratipay, since we store session tokens in the database (and only one per user). So the moment the owner of the account claims their email address in our system, the attacker's session token would be invalidated. That's assuming we don't require email address confirmation, which of course we should also do.

**Comment by whit537** [comment]
@seanlinsley Yeah, that's where I got it. Should've mentioned, sorry. :-P

**Comment by whit537** [comment]
+1 on [FJ](http://www.funnyjunk.com/Admin+changed+my+life/text/5502446/222):

>I don't use Gratipay, because you have to log in to Gratipay using Twitter, Google +, Facebook, GitHub, Bitbucket, or OpenStreetMap. I don't use any of those things and I haven't even heard of a few of them. [...] Would FJ accept donations through a different site than Gratipay?

**Comment by whit537** [comment]
Interesting. Medium has added email authentication, but without passwords. They mail you a new link every time you want to sign in.

![screen shot 2015-07-27 at 2 41 11 pm](https://cloud.githubusercontent.com/assets/134455/8914485/a2df437a-346d-11e5-964c-2937a8fee85d.png)

**Comment by whit537** [comment]
![screen shot 2015-07-27 at 2 49 09 pm](https://cloud.githubusercontent.com/assets/134455/8914690/ad257df8-346e-11e5-85eb-190a37659889.png)

**Comment by whit537** [comment]
I think we should go the Medium route. It's much simpler to implement and reduces our security exposure significantly.

**Comment by techtonik** [comment]
:+1:

**Comment by mattbk** [comment]
+1 for Persona from [FD](https://gratipay.freshdesk.com/helpdesk/tickets/3213):

>Do you plan on providing login methods that are more neutral? I'd like 
to use Personna or anything that wouldn't require me to use a 
Github/Facebook/any other non free platform.

**Comment by whit537** [comment]
I think we should make this our only authentication mechanism, and we should phase out account merging.

**Comment by whit537** [comment]
I'm putting this on "Solve Corporate Sponsorship of Open-Source" because using the company Twitter to login to Gratipay is onerous when you're sitting in one department and marketing owns Twitter and they're way over there.

**Comment by whit537** [comment]
I'm pretty sure the tokens used to authenticate during email sign-up should be treated as passwords.

https://www.owasp.org/index.php/Password_Storage_Cheat_Sheet

**Comment by whit537** [comment]
+1 from @sampablokuper in #4085.

**Comment by mattbk** [comment]
+1 from [Freshdesk](https://gratipay.freshdesk.com/helpdesk/tickets/6349):

> I understand your problem [with fraud detection]. However, a lot of people will have privacy 
issues with that. Personally, I won't sign onto anything through a 
social site because of the information that it makes available. Email 
authentication works for me, but not signing up through Facebook or Twitter.

**Comment by whit537** [comment]
+1 from Earl at the coffee shop! ☕️

**Comment by mattbk** [comment]
+1 from [FD](https://gratipay.freshdesk.com/helpdesk/tickets/6797). This particular issue is compounded by not showing giving amounts on the chalkboard when the credit card isn't working.

> So this really isn't the most helpful e-mail. There's no information here as to what my login is. I tried twitter, facebook, google, and github and none of them show me having any donations. What's the account? Seems pretty basic. I imagine most people just ignore the e-mails and drop off doing donations. Let me know how I can actually login or just remove my account.
>
> On Thu, Feb 2, 2017 at 9:09 AM, Gratipay Support <support@gratipay.com> wrote:
 Gratipay
We tried to charge your credit card $10.74 today, to fund your ongoing support for [redacted] and 5 others, but the charge failed with this message:
Declined
Fix Credit Card
Something not right? Reply to this email for help.
Sent by Gratipay, LLC | 716 Park Road, Ambridge, PA, 15003, USA
To stop receiving emails like this, change your notification preferences.

**Comment by EdOverflow** [comment]
I found the solution to this issue: https://youtu.be/VgC4b9K-gYU

**Comment by mattbk** [comment]
Would like to see #638 (2FA) as part of this.

**Comment by rohitpaulk** [comment]
Thinking about how we can split this into small pieces: 

- [ ] Split out dedicated sign-in/sign-up pages. As a side effect, this also helps solve 1) the unfriendly 401 page problem (#4489) 2) Confusion around using the wrong third-party account to sign-in, ends up creating a new account - which requires account merging to fix. 
- [ ] Allow sign-in via email (under a `beta` query param) such that any existing verified email address can be used. Things to think about: 1) Rate limiting/Captcha
- [ ] Allow sign-up via email (under a `beta` query param). Things to think about: 1) Rate limiting/Captcha 2) Claim username before email confirmation or after? 3) Participants with zero social logins - make sure no assumptions are left across codebase? 
- [ ] Remove feature flag for sign-up and sign-up via email

**Comment by whit537** [comment]
@rohitpaulk What's the purpose of separating sign-in and sign-up (apart from the side effects)?

I was envisioning a process like Medium's, where email is just another of several options under "Sign in / Sign up."

![](https://cloud.githubusercontent.com/assets/134455/26559252/7b34dac6-447c-11e7-8a57-2b30148c9b1e.gif)

**Comment by rohitpaulk** [comment]
> In terms of user experience (and apart from the side effects), what's the purpose of separating sign-in and sign-up?

By knowing the intent, we'll be able to provide better feedback. Two separate pages is just one way of addressing this, can also be done using better UI feedback on a single page. What I want to solve is "Confusion around using the wrong third-party account to sign-in, ends up creating a new account - which requires account merging to fix.". For example, we shouldn't just 'create' a new account if the user intended to sign-in using a third-party provider but chose the wrong one. 

I'll take a look at how Medium handles this, I expect that they'd place an intermediary step between registration (something like "You don't have an account, we're going to create one for you. Are you sure?") to prevent accidentally creating a new account

**Comment by whit537** [comment]
@rohitpaulk We've talked before about phasing out social logins entirely, and requiring an email for everyone on Gratipay: https://github.com/gratipay/gratipay.com/issues/3837. Do you think we should move in that direction? If so, how does that affect the steps we take here?

**Comment by rohitpaulk** [comment]
> We've talked before about phasing out social logins entirely, and requiring an email for everyone on Gratipay: #3837. Do you think we should move in that direction? If so, how does that affect the steps we take here?

That ticket is blocked on this (we can't remove other authentication methods before we introduce email), but I don't think the steps we take here must be affected by that ticket.

**Comment by rohitpaulk** [comment]
A few rough designs I'm experimenting with: 

**Dedicated Sign-in page - 1**

![screen shot 2017-06-02 at 2 40 52 pm](https://cloud.githubusercontent.com/assets/3893573/26725723/7e77c7fe-47bd-11e7-800e-d7fd380a0560.png)

**Dedicated Sign-in page - 2**

![screen shot 2017-06-02 at 2 41 03 pm](https://cloud.githubusercontent.com/assets/3893573/26725726/7e7db2f4-47bd-11e7-8f01-1d6a877f896e.png)

**Dedicated Sign-in page - 3**
![screen shot 2017-06-02 at 2 41 10 pm](https://cloud.githubusercontent.com/assets/3893573/26725725/7e7c69da-47bd-11e7-821c-c235c3081427.png)

**Sign-in modal - 1**
![screen shot 2017-06-02 at 5 55 18 pm](https://cloud.githubusercontent.com/assets/3893573/26725724/7e7a2850-47bd-11e7-9759-b44142dc8059.png)

**Sign-in modal - 2**
![screen shot 2017-06-02 at 5 55 24 pm](https://cloud.githubusercontent.com/assets/3893573/26725728/7e848e76-47bd-11e7-97b3-f1c0153f1c4a.png)

**Sign-in modal - 3**
![screen shot 2017-06-02 at 5 55 30 pm](https://cloud.githubusercontent.com/assets/3893573/26725727/7e82e832-47bd-11e7-9dd6-8668a3c4ec0f.png)

**Sign-in modal - 4**
![screen shot 2017-06-02 at 5 55 37 pm](https://cloud.githubusercontent.com/assets/3893573/26725729/7ea5de82-47bd-11e7-893b-f42a503b81fe.png)

**Comment by rohitpaulk** [comment]
(Note: The serif font is not the one we use on production, I couldn't find a downloadable version on our typography.com account)

**Comment by whit537** [comment]
Sounds good.

Re: font ... if you use `localhost:8537` it should work as that is whitelisted. Otherwise you can probably add whatever you're using instead to the whitelist on typography.com?

**Comment by rohitpaulk** [comment]
> if you use localhost:8537 it should work as that is whitelisted

These designs are from Adobe XD, not in a browser 😄. Our typography account had Ideal Sans as a downloadable font, but not Typewriter.

**Comment by whit537** [comment]
Hah! Oh. :-)

Typewriter doesn't sound familiar. Our serif is ... Chronicle, I think?

**Comment by rohitpaulk** [comment]
> Typewriter doesn't sound familiar. 

Hmm, wonder where I got that from

> Our serif is ... Chronicle, I think?

Yes, yes

**Comment by whit537** [comment]
Sign-in modal is deployed!

https://twitter.com/Gratipay/status/879365201665064961 

![xu9wtyej0b](https://user-images.githubusercontent.com/134455/27547770-2f5487d0-5a54-11e7-9030-7ff4022b0a95.gif)

**Comment by rohitpaulk** [comment]
Moving on to the next step. I'm first going to enable sign-in (not signup) via email, under a feature flag (to test). After that, will look at sign-up via email

**Comment by mattbk** [comment]
+1 from https://github.com/gratipay/gratipay.com/issues/4554#issuecomment-317133698

**Comment by whit537** [comment]
Closing in light of our decision to [shut down Gratipay](https://gratipay.news/the-end-cbfba8f50981).

Thank you all for a great run, and I'm sorry it didn't work out! 😞  💃
